### PR TITLE
[pytree] register pytree node type in both C++ pytree and Python pytree

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112278
* #112485
* __->__ #112111

Changes:

1. Add `_private_register_pytree_node` API in both C++ and Python pytree. In C++ pytree, the API will only register pytree node for C++ pytree. In Python pytree, the API will only register pytree node for Python pytree.
2. Do not allow registering a type as pytree node twice in the Python pytree.
3. Add thread lock to the Python pytree node register API.
4. The old `_register_pytree_node` API will call the `_private_register_pytree_node` API and raise a deprecation warning.
5. Add a new `register_pytree_node` API to register node type in both C++ and Python implementations.
6. Add tests to ensure a warning will be raised when the old private function is called.

cc @zou3519 @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### CODE REVIEW

File: torch/utils/_pytree.py
Comment (jon-chuang) [inline torch/utils/_pytree.py:266]: Hmm, just curious, why don't we register the cxx nodes here, is it because cxx pytree already does its own registratio?

**Comment by jon-chuang** [comment]
@XuehaiPan can we also implement registration hygeine for `torch` namespace specified in https://github.com/pytorch/pytorch/issues/112342#issuecomment-1784215755? Can do in future PR. 

Debatable whether we should allow registering under "torch" namespace from external package. Ideally no, the user should define their own namespace; this can prevent weird bugs where importing one external package can thrash the global namespace.

Is there also a way to "freeze" the namespace?

File: torch/utils/_pytree.py
Comment (XuehaiPan) [reply torch/utils/_pytree.py:266]: > is it because cxx pytree already does its own registration of the default collections?

Yes. We are going to make the C++ pytree and Python pytree analog to each other. The C++ pytree currently support:

- tuple
- list
- dict
- OrderedDict
- defaultdict
- deque
- NamedTuple
- PyStructSequence

**Comment by XuehaiPan** [comment]
> Is there also a way to "freeze" the namespace to prevent further registrations?

I think we can do this. Now, the pytree nodes defined in torch are:

```python
>>> import torch
>>> list(torch.utils._pytree.SUPPORTED_NODES)
[
    <class 'dict'>,
    <class 'list'>,
    <class 'tuple'>,
    <function namedtuple at 0x102a707c0>,
    <class 'collections.OrderedDict'>,
    <class 'torch.fx.immutable_collections.immutable_dict'>,
    <class 'torch.fx.immutable_collections.immutable_list'>,
    <class 'torch.Size'>,
    <class 'torch.return_types.*'>,
]
```

Here `torch.return_types.*` are `PyStuctSequence` types. We are going to make them out-of-box supported (like `namedtuple`s).

File: torch/_functorch/aot_autograd.py
Comment (zou3519) [inline torch/_functorch/aot_autograd.py:107]: Were these just unused? I don't see them being moved elsewhere

Comment (jon-chuang) [reply torch/_functorch/aot_autograd.py:107]: They are defined in fx immutable_collections.py

File: torch/utils/_cxx_pytree.py
Comment (zou3519) [inline torch/utils/_cxx_pytree.py:36]: As I mentioned on a previous PR: I don't think we should import python pytree in `_cxx_pytree` when optree is not available. In the future, we may have a public torch.utils.pytree that does have this behavior. But we should only do this when we're confident that the two APIs are drop-in replacements for each other, otherwise, the behavior may silently change depending on if the user has the optree package installed.

Comment (XuehaiPan) [reply torch/utils/_cxx_pytree.py]: I have moved the condition branch for `multipy` to Python pytree. Now the C++ pytree will raise an error if the dependency is not installed.

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:183]: What is the problem with deploy?

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:190]: It feels a bit weird to me that the `_cxx_pytree.register_pytree_node` and `_pytree._register_pytree_node` have different signatures.

File: torch/onnx/_internal/fx/dynamo_graph_extractor.py
Comment (zou3519) [inline torch/onnx/_internal/fx/dynamo_graph_extractor.py:47]: Why is this false?

File: torch/utils/_pytree.py
Comment (zou3519) [reply torch/utils/_pytree.py:190]: Morally, we should register everything to both python and cxx pytree and skip the registrations for cxx pytree's built-in types. One idea is that:
1. we add a "_builtin" kwarg
2. "_builtin=True" only registers to Python pytree
3. we assert that "_builtin=True" is only being called from inside torch/utils/_pytree (by looking up the stack frame)

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:158]: NB: this is technically BC-breaking, but pytree isn't technically public

File: torch/utils/_pytree.py
Comment (zou3519) [reply torch/utils/_pytree.py:271]: Can we add parity tests for all of these builtin types somehow?

**Review by zou3519 — COMMENTED** [review]
I'm a bit uncomfortable with the `register_pytree_node` APIs having different signatures. Morally it feels like they should both register to both pytree implementations, minus the special case of types that cxx pytree already knows how to handle.

File: torch/utils/_pytree.py
Comment (XuehaiPan) [reply torch/utils/_pytree.py:190]: Thanks for the suggestion and I will update the PR.

File: torch/onnx/_internal/fx/dynamo_graph_extractor.py
Comment (XuehaiPan) [reply torch/onnx/_internal/fx/dynamo_graph_extractor.py:47]: `transformers.modeling_outputs.ModelOutput` is the only type to call this function. That type is an abstract type and will never used in real use cases.

- https://github.com/huggingface/transformers/pull/26618#issuecomment-1772310393

> * `ModelOutput` is not designed to be used directly (well, to be fair, within `transformers` code base).
>   
>   * All usage should inheritate `ModelOutput` and specify the attributes an output have), see, for example, https://github.com/huggingface/transformers/blob/929134bf65ac986c12c423c30b0db8a239f3b195/src/transformers/modeling_outputs.py#L25-L47
>   * (it's kind abstract class, despite we can still use it directly, but that usage is outside `transformers` scope.)
>   * Moreover, I would say it is super rare users will use `ModelOutput` directly  (in a manually way) to store their model output (without creating a subclass first).
>     
>     * (and even more rare, this usage along with `PyTreeExtensionContext` context manager)

------

I added this `_register_cxx_pytree_node=False` because the context manager will unregister the types (currently only `ModelOutput`) when exiting the context. In the Python pytree, it pops out types in `_pytree.SUPPORTED_NODE`. But in C++ Python, we do not offer an API to delete the node registration. On the next time we enter the context manager, the C++ pytree will raise an error for duplicate registration for `ModelOutput`.

We need to discuss how we design the node registry APIs for deleting/updating the registry.

File: torch/utils/_pytree.py
Comment (XuehaiPan) [reply torch/utils/_pytree.py:183]: The `TorchpyTest.TaggingRace` test gets internal errors. I'm not sure if we can fix this by adding a thread lock here. Or we can change the import order for `torch._C`.

https://github.com/pytorch/pytorch/actions/runs/6456239485/job/17526163874?pr=109684#step:15:2360

```text
Traceback (most recent call last):
  File "<string>", line 49, in <module>
  File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/__init__.py", line 236, in <module>
    from torch._C import *  # noqa: F403
RuntimeError: get_internals: could not successfully initialize the tstate TSS key!
unknown file: Failure
C++ exception with description "r == 0 INTERNAL ASSERT FAILED at "/var/lib/jenkins/multipy/multipy/runtime/interpreter/interpreter_impl.cpp":691, please report a bug to PyTorch. 
Exception raised from newInterpreterImpl at /var/lib/jenkins/multipy/multipy/runtime/interpreter/interpreter_impl.cpp:691 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6c (0x7f39f423ee7c in /opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x84 (0x7f39f41f0174 in /opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #2: c10::detail::deprecated_AT_ASSERT() + 0 (0x7f381e541a42 in /tmp/multipy_interpreterkpkukY)
frame #3: newInterpreterImpl + 0x98 (0x7f381e53f51c in /tmp/multipy_interpreterkpkukY)
frame #4: torch::deploy::Interpreter::Interpreter(torch::deploy::InterpreterManager*, std::shared_ptr<torch::deploy::Environment>) + 0x322 (0x55706d94f3fc in ./multipy/runtime/build/test_deploy)
frame #5: torch::deploy::Interpreter::Interpreter(std::shared_ptr<torch::deploy::Environment>) + 0x4c (0x55706d95101e in ./multipy/runtime/build/test_deploy)
frame #6: void __gnu_cxx::new_allocator<torch::deploy::Interpreter>::construct<torch::deploy::Interpreter, std::shared_ptr<torch::deploy::Environment>&>(torch::deploy::Interpreter*, std::shared_ptr<torch::deploy::Environment>&) + 0x6d (0x55706d954b05 in ./multipy/runtime/build/test_deploy)
frame #7: void std::allocator_traits<std::allocator<torch::deploy::Interpreter> >::construct<torch::deploy::Interpreter, std::shared_ptr<torch::deploy::Environment>&>(std::allocator<torch::deploy::Interpreter>&, torch::deploy::Interpreter*, std::shared_ptr<torch::deploy::Environment>&) + 0x3a (0x55706d953520 in ./multipy/runtime/build/test_deploy)
frame #8: torch::deploy::Interpreter& std::vector<torch::deploy::Interpreter, std::allocator<torch::deploy::Interpreter> >::emplace_back<std::shared_ptr<torch::deploy::Environment>&>(std::shared_ptr<torch::deploy::Environment>&) + 0x50 (0x55706d95216c in ./multipy/runtime/build/test_deploy)
frame #9: torch::deploy::InterpreterManager::InterpreterManager(unsigned long, std::shared_ptr<torch::deploy::Environment>) + 0x1d9 (0x55706d94ca29 in ./multipy/runtime/build/test_deploy)
frame #10: TorchpyTest_TaggingRace_Test::TestBody() + 0x7c (0x55706d8cec5c in ./multipy/runtime/build/test_deploy)
frame #11: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x69 (0x55706d93a566 in ./multipy/runtime/build/test_deploy)
frame #12: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x5e (0x55706d932ae7 in ./multipy/runtime/build/test_deploy)
frame #13: testing::Test::Run() + 0xf2 (0x55706d90acb4 in ./multipy/runtime/build/test_deploy)
frame #14: testing::TestInfo::Run() + 0x113 (0x55706d90b6c9 in ./multipy/runtime/build/test_deploy)
frame #15: testing::TestSuite::Run() + 0x131 (0x55706d90bfcb in ./multipy/runtime/build/test_deploy)
frame #16: testing::internal::UnitTestImpl::RunAllTests() + 0x431 (0x55706d918e55 in ./multipy/runtime/build/test_deploy)
frame #17: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x69 (0x55706d93b627 in ./multipy/runtime/build/test_deploy)
frame #18: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 0x5e (0x55706d933dad in ./multipy/runtime/build/test_deploy)
frame #19: testing::UnitTest::Run() + 0xcd (0x55706d91756f in ./multipy/runtime/build/test_deploy)
frame #20: RUN_ALL_TESTS() + 0x15 (0x55706d8d93f0 in ./multipy/runtime/build/test_deploy)
frame #21: main + 0x9d (0x55706d8d348d in ./multipy/runtime/build/test_deploy)
frame #22: __libc_start_main + 0xf3 (0x7f39e5779083 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #23: _start + 0x2e (0x55706d8c855e in ./multipy/runtime/build/test_deploy)
" thrown in the test body.
[  FAILED  ] TorchpyTest.TaggingRace (14370 ms)
```

https://github.com/pytorch/multipy/blob/765dfef74b385532a2f707c20af8407d7f91cf70/multipy/runtime/interpreter/interpreter_impl.cpp#L690-L691

File: torch/utils/_pytree.py
Comment (XuehaiPan) [reply torch/utils/_pytree.py:271]: > Can we add parity tests for all of these builtin types somehow?

We can do this since we do not allow registering a single type twice.

```python
for cls in builtin_types:
    with assertRaises(TypeError):
        pytree.register_pytree_node(cls, ...)
```

I will add these tests after we make the builtin types for Python pytree aligned with C++ pytree. I'll open follow-up PRs to:

- add defaultdict and deque
- add function `is_structseq` / `is_stuctseq_class` and support `torch.return_types.*` natively.

File: torch/utils/_pytree.py
Comment (XuehaiPan) [reply torch/utils/_pytree.py]: Further discussion moved to PR:

- #112477

I add a warning that C++ pytree is not available under `torchdeploy`.

File: torch/utils/_pytree.py
Comment (XuehaiPan) [reply torch/utils/_pytree.py]: I add a PR to ensure the two implementations provides the same public APIs with the same function signatures.

- #112485

File: torch/_export/utils.py
Comment (zou3519) [inline torch/_export/utils.py:37]: I'm confused, is there overlap between `__qualname__` and `__module__`?

File: torch/utils/_cxx_pytree.py
Comment (zou3519) [inline torch/utils/_cxx_pytree.py:35]: We should probably just raise here instead of warn?

File: torch/utils/_cxx_pytree.py
Comment (zou3519) [inline torch/utils/_cxx_pytree.py:232]: This seems a bit dodgy. Can we get it to work without inspect? It seems like we should have a private `_private_register_pytree_node` APIs in both cxx_pytree and python_pytree and then have `_register_pytree_node` in both APIs call `cxx_pytree._private_register_pytree_node` and `python_pytree._private_register_pytree_node`. We can prevent people from using both of these directly by stating in the docstring what they should use.

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:185]: - Importing cxx_pytree should fail when running_with_deploy()
- Then, we don't need to warn here: users will see the warning if they are using deploy, and importing cxx_pytree.

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:191]: It would be nice if there was some way to tell that cxx pytree is available or not. Something like:
```
import torch.utils._cxx_pytree
torch.utils._cxx_pytree.is_available()
```
In this regime, we would probably need to get the cxx_pytree APIs to error if cxx_pytree is not available

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:273]: Probably should use a `python_pytree._private_register_pytree_node` helper function  (mentioned above) to avoid repetition

**Review by zou3519 — COMMENTED** [review]
We should probably have separate "private" `_private_register_pytree_node` APIs, one for C++, and one for Python. Then, we'd tell users to actually use the `_register_pytree_node` APIs, which will call into both. This should clean up some of the logic. What do you think?

File: torch/utils/_cxx_pytree.py
Comment (XuehaiPan) [reply torch/utils/_cxx_pytree.py:232]: I have updated the PR.

File: torch/_export/utils.py
Comment (XuehaiPan) [reply torch/_export/utils.py:37]: > is there overlap between `__qualname__` and `__module__`?

No. The `__qualname__` refers to the fully qualified name. The `__qualname__` is different than `__name__` only in nested cases. Otherwise, they are the same.

[PEP 3155 – Qualified name for classes and functions](https://peps.python.org/pep-3155/#example-with-nested-classes)

**Comment by XuehaiPan** [comment]
> We should probably have separate "private" `_private_register_pytree_node` APIs, one for C++, and one for Python.

I have updated the PR accordingly.

**Review by zou3519 — DISMISSED** [review]

**Comment by zou3519** [comment]
We should take a look at the callsites. It's possible that this is more BC-breaking than we expected.

**Comment by zou3519** [comment]
Sorry, I was out sick last week. I think what I wanted to do with this PR was:
1) determine if it is too BC-breaking
2) If so, change the behavior

For 2, some ideas are:
- we can warn if we see a duplicate registration
- we can allow the duplicate registration through (can we do this somehow for C++ pytree too?)

**Comment by zou3519** [comment]
Okay, here's a more well-formed idea:
1) we should create a new register_pytree_node, that has the "correct" behavior. That is, it will error out on duplicate registrations and it will also register to both Python and C++ pytree.
2) We should keep around the old _register_pytree_node. This one should not error on duplicate registrations and also will not register to both Python/C++ pytree. It should emit a deprecation warning
3) We're going to want to change all, if not most, things in PyTorch to use the new register_pytree_node.
4) Internal things: I'll drive those migrations

Thoughts?

**Comment by XuehaiPan** [comment]
> Okay, here's a more well-formed idea:
> 
> 1. we should create a new register_pytree_node, that has the "correct" behavior. That is, it will error out on duplicate registrations and it will also register to both Python and C++ pytree.
> 2. We should keep around the old _register_pytree_node. This one should not error on duplicate registrations and also will not register to both Python/C++ pytree. It should emit a deprecation warning
> 3. We're going to want to change all, if not most, things in PyTorch to use the new register_pytree_node.
> 4. Internal things: I'll drive those migrations
> 
> Thoughts?

@zou3519 I agree with these. I can do steps 1-2 in this PR.

**Comment by XuehaiPan** [comment]
I have updated the PR accordingly.

**Comment by zou3519** [comment]
Can you update the PR description?

File: torch/_export/utils.py
Comment (zou3519) [inline torch/_export/utils.py:125]: We should probably leave this one alone? I'm not sure if this causes internal problems. We can swap it over later.

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:217]: Can we add a test for the warning? Also the path is torch.utils._pytree.register_pytree_node

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:137]: Let's delete to_str_fn and maybe_from_str_fn for the torch.utils._pytree.register_pytree_node API, since those are deprecated and we are creating a new api here.

**Review by zou3519 — COMMENTED** [review]
Looks mostly good to me. Since we're creating a "new" register_pytree_node API we can delete the deprecated args from it. Also, let's add a test for a warning, I'm worried pytorch might hide warnings by default but I don't remember

**Comment by XuehaiPan** [comment]
I have made the requested changes in the code, tests, and PR description.

**Review by zou3519 — APPROVED** [review]

**Comment by DanilBaibak** [comment]
Hi @XuehaiPan! Sorry, I need to revert your PR because it breaks the internal build. I'm working on a fix. As soon as fix is ready, we will merge your changes back.

**Comment by XuehaiPan** [comment]
> Sorry, I need to revert your PR because it breaks the internal build. I'm working on a fix.

Hi @DanilBaibak, is there anything I can help?

**Comment by DanilBaibak** [comment]
Hi @XuehaiPan! It looks only like internal changes are required. I have already a fix. As soon as I get approval, I will merge your PR.

File: torch/utils/_pytree.py
Comment (zou3519) [inline torch/utils/_pytree.py:125]: @XuehaiPan, I'm reading the code again. `register_pytree_node` should error out on duplicate registrations. However, `_register_pytree_node` should not error out on duplicate registrations, to maintain BC. This doesn't appear to be the case, and this is what is tripping the internal tests.

**Review by zou3519 — APPROVED** [review]
thank you! Hopefully this PR sticks this time.

**Comment by XuehaiPan** [comment]
> I have already a fix. As soon as I get approval, I will merge your PR.

Hi @DanilBaibak, I updated the PR to ensure the old APIs are backward compatible. Once you get ready, you can trigger a merge.

**Pull request closed without merging**
